### PR TITLE
Meshlet continuous LOD

### PR DESCRIPTION
**Pull request by JMS55**
Adds a basic level of detail system to meshlets. An extremely brief summary is as follows:
* In `from_mesh.rs`, once we've built the first level of clusters, we group clusters, simplify the new mega-clusters, and then split the simplified groups back into regular sized clusters. Repeat several times (ideally until you can't anymore). This forms a directed acyclic graph (DAG), where the children are the meshlets from the previous level, and the parents are the more simplified versions of their children. The leaf nodes are meshlets formed from the original mesh.
* In `cull_meshlets.wgsl`, each cluster selects whether to render or not based on the LOD bounding sphere (different than the culling bounding sphere) of the current meshlet, the LOD bounding sphere of its parent (the meshlet group from simplification), and the simplification error relative to its children of both the current meshlet and its parent meshlet. This kind of breaks two pass occlusion culling, which will be fixed in a future PR by using an HZB from the previous frame to get the initial list of occluders.

Many, _many_ improvements to be done in the future https://github.com/bevyengine/bevy/issues/11518, not least of which is code quality and speed. I don't even expect this to work on many types of input meshes. This is just a basic implementation/draft for collaboration. 

Arguable how much we want to do in this PR, I'll leave that up to maintainers. I've erred on the side of "as basic as possible". 

References:
* Slides 27-77 (video available on youtube) https://advances.realtimerendering.com/s2021/Karis_Nanite_SIGGRAPH_Advances_2021_final.pdf 
* https://blog.traverseresearch.nl/creating-a-directed-acyclic-graph-from-a-mesh-1329e57286e5
* https://jglrxavpok.github.io/2024/01/19/recreating-nanite-lod-generation.html,  https://jglrxavpok.github.io/2024/03/12/recreating-nanite-faster-lod-generation.html, https://jglrxavpok.github.io/2024/04/02/recreating-nanite-runtime-lod-selection.html, and https://github.com/jglrxavpok/Carrot
* https://github.com/gents83/INOX/tree/master/crates/plugins/binarizer/src
* https://cs418.cs.illinois.edu/website/text/nanite.html

image
image

### CODE REVIEW

**Review by atlv24 — APPROVED** [review]
Looks good, I'd probably rename `self_lod` and `self_culling` to just `lod` and `culling`

**Comment by JMS55** [comment]
> Looks good, I'd probably rename self_lod and self_culling to just lod and culling

I'd personally like to keep it extra explicit, because it's really easy to get confused :sweat_smile:.

File: crates/bevy_pbr/src/lib.rs
Comment (mockersf) [inline crates/bevy_pbr/src/lib.rs:4]: could you remove this from the crate level and allow it at the usage location?

Comment (JMS55) [reply crates/bevy_pbr/src/lib.rs:4]: I tried leaving the forbid in place and putting an allow on the specific file, but it didn't seem to like that. What's the correct way to do this?

Comment (pcwalton) [reply crates/bevy_pbr/src/lib.rs:4]: You'd have to switch it to `#[deny]`; then `#[allow]` would work. In general I don't prefer using `#[forbid]` as it hurts debuggability for questionable gain--tools can easily find unsafe code without or without it.

Comment (JMS55) [reply crates/bevy_pbr/src/lib.rs]: Fixed, thanks.

File: crates/bevy_pbr/src/meshlet/asset.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/asset.rs:136]: Maybe this could just be `MeshletSaveOrLoadError` to make it slightly less wordy? Are there any other assets that meshlets might want to load and save?

(I was originally going to suggest `MeshletIoError`, but `MeshletIoError::Io()` would be confusing, so yeah)

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:112]:
```suggestion
    let sphere_diameter_uv = view.projection[0][0] * r * inverseSqrt(d2 - r2);
```

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:91]: At some point we might want to consider importing `itertools`, as using `max` from that crate would be easier to read.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:182]:
```suggestion
    for meshlet in &mut meshlets.meshlets {
```

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:192]: Can we add a safe wrapper for this to the `meshopt` crate instead? It looks like the safe wrapper would be trivial, and that would let us avoid messing with the unsafe code settings in this crate.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:253]: I think I'd prefer some more descriptive names here. Also a comment explaining what's going on, because the algorithm isn't obvious to me :)

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:364]:
```suggestion
    #[error("Mesh attributes are not {{POSITION, NORMAL, UV_0, TANGENT}}")]
```

File: crates/bevy_pbr/src/meshlet/asset.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/asset.rs:33]: Maybe `max_meshlet_triangles`?

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:300]: I personally like factoring tunable parameters like this out into `const`s at the top of the file so people can play with them if they need to. Up to you.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:315]: Same here.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:266]: Why 4?

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:310]: So we aren't doing the Nanite thing where they overlap LODs to avoid locking borders, right? We're just locking borders for now?

(I agree with your approach for simplicity's sake, because I never understood what Nanite does. Just wanting to understand)

**Review by pcwalton — COMMENTED** [review]
I did a look over this, mostly for code style and so forth, not so much for the algorithm. In general this looks pretty straightforward. I definitely appreciate the clarity from the comments and splitting this up into a smaller patch!

Exciting stuff :)

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:112]: The spec says inverseSqrt() does not work if e <= 0. Need to think if this is an issue.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:91]: I already use itertools elsewhere in this file, I'll use it.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:191]: Yeah it's not published yet though, so for now, unsafe it is.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:252]: METIS takes a very specific input format. The names are copied from what METIS expects. I linked the METIS manual, read that to understand how it works. Comments wouldn't really help.

File: crates/bevy_pbr/src/meshlet/asset.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/asset.rs:33]: I personally feel that that's less clear. Up to maintainers.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:299]: I used to have it in a previous commit. I can't decide why I removed it. Probably because it's just a heuristic that mostly works for the meshes I tested. Might need to be tunable by users in the future, idk. It'll need further testing when we try it out on a large array of meshes, but this PR was just to get something usable.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:265]: 4 meshlets per group, ideally. 4 meshlets (256 triangles), simplify group to 50% (128 triangles), split back into meshlets = 2 meshlets (64 x 2 triangles). 

I should probably rename this to target_partition_count, there are no hard limits.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:309]: No, we are doing what nanite does. What you want to do is avoid cracks/gaps when displaying different LOD levels for different sections of a mesh.

By grouping meshlets, locking the group border, simplifying the group, and then splitting into new meshlets to form a LOD, you avoid the cracks between LOD levels.

I can't really explain why that works intuitively, but read the nanite slides or the PhD thesis that nanite is based on (link in slides) if you want the theory.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:91]: Won't work on f32.

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl
Comment (atlv24) [reply crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:112]: i dont think sqrt is guaranteed to work for e <= 0 either. I'd just slab an abs or max 0 on it if its a worry, being explicit about edge case behavior intent is valuable and saturating/clamping is often free in gpus

**Review by atlv24 — APPROVED** [review]

File: crates/bevy_pbr/src/meshlet/asset.rs
Comment (atlv24) [reply crates/bevy_pbr/src/meshlet/asset.rs:33]: i prefer keeping it as is

File: assets/models/bunny.meshlet_mesh
Comment (superdump) [inline assets/models/bunny.meshlet_mesh:1]: Do you think you will need to change this more? Storing binary files in git and having history of them isn't the nicest thing so if we could avoid it, that would be good. :)

Comment (JMS55) [reply assets/models/bunny.meshlet_mesh:1]: Yep, it will definitely change a couple more times... Storing it externally also sucks though as then you can't run the example easily... Idk what we should do.

Comment (atlv24) [reply assets/models/bunny.meshlet_mesh:1]: It will almost certainly change again, the format is not compressed yet.

Comment (JMS55) [reply assets/models/bunny.meshlet_mesh:1]: If I get Gltf->MeshletMesh asset preprocessing in, we can ship the repo with the GLTF file, and just have users convert it as part of the example. That way we won't have to worry about the MeshletMesh asset changing. But that's a good bit off, I don't want to block on that.

Comment (superdump) [reply assets/models/bunny.meshlet_mesh:1]: @bevyengine/maintainer-team any thoughts on this?

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (superdump) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:61]: This feels very much like it could be a for loop, which would feel less error prone.

Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:61]: It's not a for loop. We dynamically adjust the queue range to process more items at the end of the loop.

Iter 1: 0..1000
Iter 2: 1000..1500
Iter 3: 1500..1700
...

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (superdump) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:252]: Could a link to the relevant section of METIS be added here?

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (superdump) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:265]: Earlier code mentions that the meshlets are grouped into _roughly_ 4 in a group. How come 4 is always used here if there aren't guaranteed to be 4?

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (superdump) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:131]: One thing I don't understand about this loop is why each iteration processes the whole range from current lod to max lod - why doesn't it just process for the current lod from the previous lod?

**Review by superdump — COMMENTED** [review]
Just a few nits/clarifications, otherwise this looks good to me.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (superdump) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:61]: Ahhh, meshlets is mutated in split_simplified_groups_into_new_meshlets. I see.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:265]: It's the partition count. Given N nodes (meshlets), Metis divides them into P (partition_count) partitions (groups).

So given 1000 meshlets, we tell Metis to divide the 1000 meshlets into (1000/25)=250 groups. It's not guaranteed that we have _exactly_ 4 per group though. Some groups might have slightly more or less than 4 meshlets, as it tries to spread the cost out as defined by the edge weights between nodes. We just have to hope for an even split between all partitions, which would give 4 meshlets per group if that were to happen.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (superdump) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:131]: I had misunderstood. I get it now.

**Review by superdump — APPROVED** [review]
I approve, but I still want to hear other maintainers' thoughts about the meshlet mesh file. Every change will go into the history and everyone not doing a shallow clone, which will be basically everyone, will download the whole history of all files. That isn't good.

File: assets/models/bunny.meshlet_mesh
Comment (superdump) [reply assets/models/bunny.meshlet_mesh:1]: Just to note as it happened on Discord - it was decided that hosting the file out of tree would be better, and prompting the user to download it. The reason is that binary files that are modified and committed to the repository will have all versions in git history and the only way to remove them is to rewrite the git history.”, which isn’t nice. But also, rewriting git history breaks all open PRs and branches.

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1303 if you'd like to help out.

**Pull request merged**
